### PR TITLE
Move balance check into debug assertion

### PR DESCRIPTION
**Pull request by mooori**
It was decided offline that `check_balance` should not be called in release builds anymore. This PR implements this change by gating `check_balance(...)` behind `cfg!(debug_assertions)`.

### CODE REVIEW

**Review by akhi3030 — APPROVED** [review]

File: runtime/runtime/src/lib.rs [hunk 1]
```diff
@@ -2084,16 +2084,20 @@ impl Runtime {
...
+                &apply_state.config,
+                &state_update,
+                validator_accounts_update,
+                processing_state.incoming_receipts,
+                &promise_yield_result.timeout_receipts,
+                processing_state.transactions,
+                &receipt_sink.outgoing_receipts(),
+                &processing_state.stats,
+            ),
+            Ok(()),
```
Comment (shreyan-gupta) [inline runtime/runtime/src/lib.rs:2098]: Instead of comparing with `Ok(())` should we consider unwrapping the error here for better tracing?

Comment (mooori) [reply runtime/runtime/src/lib.rs:2098]: `assert_eq!` prints `left` and `right` on a mismatch, see for instance [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=28eb202c0101fbacc7b747d15aa78b01). Though I might rewrite it to add tracing.

**Comment by mooori** [comment]
I've rewritten it to panic with the error message plus some context. That feels like the closest match to `debug_assert` with logging / tracing.

**Review by tayfunelmas — APPROVED** [review]

File: runtime/runtime/src/lib.rs [hunk 2]
```diff
@@ -2084,16 +2084,23 @@ impl Runtime {
...
+                &state_update,
+                validator_accounts_update,
+                processing_state.incoming_receipts,
+                &promise_yield_result.timeout_receipts,
+                processing_state.transactions,
+                &receipt_sink.outgoing_receipts(),
+                &processing_state.stats,
+            ) {
+                panic!(
+                    "Balance check failed for shard {} at height {} with block hash {}: {}",
```
Comment (tayfunelmas) [inline runtime/runtime/src/lib.rs:2099]: maybe add more context here, because it is not clear to me what you mean by "balance check".

**Review by akhi3030 — APPROVED** [review]

**Comment by akhi3030** [comment]
I think this is good to be merged.  Any further suggestions can be made in additional PRs.

**Pull request merged**